File: ParseGenotypeData/convert.pl
#!/usr/bin/perl
#
# Combine the data in "genotypes.txt", "markers.txt" and "families.txt" 
# and convert into a CRI-MAP .gen file.

# file names
$gfile = "genotypes.txt";  # genotype data
$mfile = "markers.txt";    # list of markers, in order
$ffile = "families.txt";   # family information 
$ofile = "data.gen";       # output file

# read marker names and place in the vector @markers
open(IN, $mfile) or die("Cannot read from $mfile");
while($line = <IN>) {
    chomp($line);
    push(@markers, $line);
}
close(IN);

# read family information and place in the hashes %dad, %mom, %sex
open(IN, $ffile) or die("Cannot read from $ffile");
$line = <IN>;
while($line = <IN>) {
    chomp($line);
    @v = split(/\s+/, $line);
    if($v[0] eq "") { shift @v; }
    
    ($fam, $ind, $dad, $mom, $sex) = @v;
    
    $dad{$fam}{$ind} = $dad;
    $mom{$fam}{$ind} = $mom;

    if($sex == 2) {   # make female = 0 rather than = 2 
	$sex{$fam}{$ind} = 0;
    }
    else {  
	$sex{$fam}{$ind} = $sex;
    }
}
close(IN);

# read genotype data
open(IN, $gfile) or die("Cannot read from $gfile");

# header line : parse family/individual IDs
$line = <IN>; chomp($line);
@v = split(/\s+/, $line);
if($v[0] eq "") { shift @v; } # if first entry is blank, get rid of it
shift @v; # get rid of the first entry "Marker"
foreach $v (@v) {
    ($fam,$ind) = split(/-/, $v); # split at '-'
    @fam = (@fam, $fam);
    @ind = (@ind, $ind);
}

# parse rest of file 
while($line = <IN>) {

    $marker = substr($line, 0, 9);
    $marker =~ s/\s+//g; # remove any extra spaces
    
    foreach $i (0..(@fam-1)) {
	$g = substr($line, 9+$i*7, 7); 
	($g1, $g2) = split(/\//, $g);  # split at '/'
	$g1 =~ s/\s+//g; # remove any extra spaces
	$g2 =~ s/\s+//g;

	if($g1 == 0 or $g2 == 0) { # replace blanks with 0's
	    $g1 = 0;
	    $g2 = 0;
	}

	$gen{$fam[$i]}{$ind[$i]}{$marker} = $g1 . " " . $g2;
	
    }
}
close(IN);


# write genotype data as cri-map .gen file
open(OUT, ">$ofile") or die("Cannot write to $ofile");

$nfam = keys %gen;      # number of families
print OUT ("$nfam\n");
$nmar = @markers;       # number of markers
print OUT ("$nmar\n");
foreach $mar (@markers) {
    print OUT ("$mar\n");
}

foreach $fam (sort numerically keys %gen) {  # families in numerical order
    print OUT ("$fam\n");
    $nind = keys %{$gen{$fam}};  # number of individuals in family
    print OUT ("$nind\n");

    foreach $ind (sort numerically keys %{$gen{$fam}}) {
	print OUT ("$ind $mom{$fam}{$ind} $dad{$fam}{$ind} $sex{$fam}{$ind}\n");
	
	foreach $mar (@markers) {  # markers in order
	    print OUT ("$gen{$fam}{$ind}{$mar} ");
	}

	print OUT ("\n");
    }
}
close(OUT);


# subroutine to allow sorts by numerical order
sub numerically { $a <=> $b; }

File: ParseGenotypeData/convert.py
#!/usr/bin/env python3
# Combine the data in "genotypes.txt", "markers.txt" and "families.txt"
# and convert them into a CRI-MAP .gen file
#
# This is a python version of the perl script discussed in my "Intro to perl"
# at http://www.biostat.wisc.edu/~kbroman/perlintro/index.html#ex2
#
# This script requires python3

import re

def read_markers (filename):
  lines = open(filename, 'r').readlines()
  return [line.strip() for line in lines]

class Person:
  def __init__ (self,family, id, dad, mom, sex):
    self.family = family
    self.id = id
    self.dad = dad
    self.mom = mom
    self.sex = "0" if sex == "2" else sex # convert from 1/2 to 1/0
    self.famid = family + '-' + id
    self.gen = {}

# read family info and return a hash of people
def read_families (filename):
  file = open(filename, 'r')
  file.readline() # header row
  people = {}
  for line in file:
    vals = line.strip().split()    # strip removes leading and ending white space
    person = Person(vals[0], vals[1], vals[2], vals[3], vals[4])
    people[person.famid] = person
  return people

# clean up string - > genotype
def parse_genotype (string):
  string = re.sub(r'\s+', '', string)
  string = "0/0" if string == "" else string
  return re.sub(r'/', ' ', string)

# read genotype data, fill in genotypes within people hash
def read_genotypes (filename, people):
  file = open(filename, 'r')

  header = file.readline().strip().split()
  header = header[1:] # omit the first field, "Marker"

  for line in file:
    marker = re.sub(r'\s+', '', line[:9])
    line = line[9:]
    for i in range(len(header)):
      person = header[i]
      start = i*7
      people[person].gen[marker] = parse_genotype(line[start:(start+7)])

# distinct families
def get_families (people):
  return set([people[key].family for key in people])

# people within a family
def get_family_members (people, family):
  return [people[key] for key in people if people[key].family == family]

def write_genfile (filename, people, markers):
  file = open(filename, 'w')
  
  families = sorted(get_families(people))
  print(len(families), file=file)

  print(len(markers), file=file)
  for marker in markers:
    print(marker, file=file)

  for family in families:
    print(family, file=file)
    members = sorted(get_family_members(people, family), key=lambda person: int(person.id))
    print(len(members), file=file)

    for person in members:
      print("%s %s %s %s" % (person.id, person.mom, person.dad, person.sex), file=file)

      for marker in markers:
        print(person.gen[marker], end=" ", file=file)
      print(file=file)


if __name__ == '__main__':
  # file names
  gfile = "genotypes.txt" # genotype data
  mfile = "markers.txt"   # list of markers, in order
  ffile = "families.txt"  # family information
  ofile = "data.gen"      # output file

  # read the data
  markers = read_markers(mfile)
  people = read_families(ffile)
  read_genotypes(gfile, people)

  # write the data
  write_genfile(ofile, people, markers)
